### PR TITLE
[query] Add module builder

### PR DESCRIPTION
**Pull request by cseed**
Stacked on: https://github.com/hail-is/hail/pull/8179

This adds ModuleBuilder.  A module is a collection of classes whose bytecode should be loaded together (like a function and some associated dependent functions).  Dependent functions now add themselves to the module.

### CODE REVIEW

File: hail/src/main/scala/is/hail/asm4s/FunctionBuilder.scala [hunk 1]
```diff
@@ -27,7 +27,59 @@ class Field[T: TypeInfo](classBuilder: ClassBuilder[_], val name: String) {
...
+
+    }
+    classesBytes
+  }
+}
+
+class ClassBuilder[C](
+  module: ModuleBuilder,
+  val name: String = null) {
+  // FIXME use newClass
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/asm4s/FunctionBuilder.scala]: Is this a `FIXME` before merging or one for sometime in the future?

Comment (johnc1231) [inline hail/src/main/scala/is/hail/asm4s/FunctionBuilder.scala]: Should this be fixed before merging or later?

Comment (cseed) [reply hail/src/main/scala/is/hail/asm4s/FunctionBuilder.scala]: Let's say sometime in the future.

Comment (cseed) [reply hail/src/main/scala/is/hail/asm4s/FunctionBuilder.scala]: I'm going to be making a bunch more passes over this code.

**Pull request merged**